### DISCUSSION TITLE
How to use Timezones output?

### DISCUSSION THREAD

**Discussion by s22-tech** [YOURLS/YOURLS#3196] [Plugins]
How can we use the output from the [timezones](https://github.com/YOURLS/timezones) plugin in our own plugins?  I'd like to get away from having to add date_default_timezone_set() in my plugins to get the correct time.

Thanks,
Marc

**Comment by s22-tech** [comment]
I found it:

```
$tz = yourls_get_option('timezone');
date_default_timezone_set($tz['time_zone']);
```
I was under the impression that custom plugins would automatically use the timezone from Yourls, but apparently Yourls is not pulling this information yet since Timezones is an optional plugin itself?

**Reply by ozh** [reply]
Yes, timezone is an optional plugin for people who care about time zones. Many don't :)

You should use YOURLS core functions ([see](https://github.com/YOURLS/YOURLS/blob/master/includes/functions-formatting.php#L734-L790)) to get time or date. This way, if the Timezone plugin is activated, you'll get timezoned times and dates, and if not you'll get user defined offset time.

**Reply by s22-tech** [reply]
I'm just trying to set a variable to the current time in my timezone but none of the functions work for me so I must be doing something wrong.  I'm guessing that `yourls_get_time_format($format)` would be the one I need in this case, but I can't find any examples of what to use for $format.  I want to mimic `date('g:i a')` but using `yourls_get_time_format('g:i a')` prints H:i.

**Reply by s22-tech** [reply]
I just found it.  If anyone else is stuck on this, here's how to do it:
`yourls_date_i18n(yourls_get_time_format('g:i a'))`

**Reply by s22-tech** [reply]
Except it's not printing AM or PM.  What am I missing?

**Reply by s22-tech** [reply]
Something appears to be going on with the formatting functions.  I also tried this:
```
yourls_date_i18n(yourls_get_date_format('Y-m-d'))
```
and it prints out `2022/03/15` instead of `2022-03-15`.

**Comment by s22-tech** [comment]
The formatting results I mentioned in my responses above seem to be governed by the defaults chosen in /timezones/admin.php
I added `'Y-m-d',  // 2022-03-15` to the $choices array, selected that as my default, and the date now prints in that format in my plugin.  Same situation with the time format.

So, how can we choose a different format to show in a plugin than we have selected in the admin?  I would think we should be able to choose the formatting we want to use in different places.  Isn't that what `yourls_get_date_format($format)` is for?

**Reply by ozh** [reply]
The plugin defines the default format. If you want to use another format, just use another format :)

```php
$now = time();
var_dump($now); // int(1647789745)

// 'M d, Y H:i' or a custom format if specified with the plugin :
var_dump(yourls_date_i18n( yourls_get_datetime_format('M d, Y H:i'), yourls_get_timestamp( $now ))); // string(16) "2022/03/20 16:22"

// always and only 'M d, Y H:i' 
var_dump(yourls_date_i18n( 'M d, Y H:i', yourls_get_timestamp( $now ))); // string(18) "Mar 20, 2022 16:22"
```

**Reply by s22-tech** [reply]
Thanks, Ozh.  That all makes sense except that only your last example works for me.  No matter which format I use with yourls_get_datetime_format(), it only prints in the format selected in the timezone plugin.  Something is keeping me from overriding it - unless, of course, that's what your comment "if specified with the plugin" means.  

And if that's the case, what's the purpose of yourls_get_datetime_format() in this context?

**Reply by ozh** [reply]
See [source](https://github.com/YOURLS/YOURLS/blob/3828863675ce8257704e402ac4d9f7c8ee396168/includes/functions-formatting.php#L764-L773). `yourls_get_datetime_format('something')` returns `'something'`, but it's filterable so it can be altered by a filter.
